core/src/main/scala/tech/beshu/ror/accesscontrol/domain/userAndGroups.scala

```diff
@@ -138,32 +144,42 @@ object PermittedGroupIds {
     }
 
     def matches(groupId: GroupId): Boolean = {
-      permittedGroupIds.matcher.`match`(groupId)
+      groupIds.matcher.`match`(groupId)
     }
   }
 }
 
-final case class ResolvablePermittedGroupIds(permittedGroupIds: UniqueNonEmptyList[RuntimeMultiResolvableVariable[GroupIdLike]]) {
-  def resolveGroupIds[B <: BlockContext](blockContext: B): Option[PermittedGroupIds] = {
-    UniqueNonEmptyList
-      .from(resolveAll(permittedGroupIds.toNonEmptyList, blockContext))
-      .map(PermittedGroupIds.apply)
+sealed trait GroupsLogic
+
+object GroupsLogic {
+
+  sealed trait PositiveGroupsLogic extends GroupsLogic {
+    val permittedGroupIds: GroupIds
   }
-}
 
-sealed trait GroupsLogic {
-  val permittedGroupIds: PermittedGroupIds
-}
+  final case class Or(override val permittedGroupIds: GroupIds) extends PositiveGroupsLogic
 
-object GroupsLogic {
-  final case class Or(override val permittedGroupIds: PermittedGroupIds) extends GroupsLogic
-  final case class And(override val permittedGroupIds: PermittedGroupIds) extends GroupsLogic
+  final case class And(override val permittedGroupIds: GroupIds) extends PositiveGroupsLogic
+
+  sealed trait NegativeGroupsLogic extends GroupsLogic {
+    val forbiddenGroupIds: GroupIds
+  }
+
+  final case class NotAnyOf(override val forbiddenGroupIds: GroupIds) extends NegativeGroupsLogic
+
+  final case class NotAllOf(override val forbiddenGroupIds: GroupIds) extends NegativeGroupsLogic
+
+  final case class CombinedGroupsLogic(positiveGroupsLogic: PositiveGroupsLogic,
+                                       negativeGroupsLogic: NegativeGroupsLogic) extends GroupsLogic
```

Comment:
Let's add some tests for CombinedGroupsLogic usage within AuthorizationRules. Currently, only the settings parsing is tested.

Reply:
Added tests. One invalid situation was detected, I modified implementation for CombinedGroupsLogic 

Also, I extracted common decoder for GroupsLogic for all Auth rules.